packages/engine/src/worker/runner/javascript/data_ffi.rs

```diff
@@ -0,0 +1,119 @@
+use std::{self, ptr::NonNull};
+
+use super::{new_js_string, Error, Result};
+
+/// C representation of Arrow array data nodes
+#[repr(C)]
+#[derive(Debug)]
+pub struct DataFfi<'s> {
+    pub len: usize,
+    pub null_count: usize,
+    pub n_buffers: u32,
+    // The number of valid pointers in buffer_ptrs and valid capacities in buffer_capacities is
+    // equal to n_buffers, which is at most 2.
+    pub buffer_ptrs: [*const u8; 2],
+    pub buffer_capacities: [usize; 2],
+    // This pointer can be null if the Arrow array doesn't have a null bitmap
+    pub null_bits_ptr: *const u8,
+    pub null_bits_capacity: usize,
+    pub(crate) _phantom: std::marker::PhantomData<v8::Local<'s, ()>>,
+}
+
+impl<'s> DataFfi<'s> {
+    pub(crate) fn new_from_js(
+        scope: &mut v8::HandleScope<'s>,
+        data: v8::Local<'s, v8::Value>,
+    ) -> Result<DataFfi<'s>> {
+        let obj = data
+            .to_object(scope)
+            .ok_or_else(|| Error::V8("Could not convert data from Value to Object".to_string()))?;
+
+        let len_value = get_obj_property(scope, obj, "len")?;
+        let len_num: v8::Local<'s, v8::Number> = len_value.try_into().map_err(|err| {
+            Error::V8(format!(
+                "Could not convert len_value from Value to Number: {err}"
+            ))
+        })?;
```

Comment:
Can we avoid this code duplication?

Reply:
Yes https://github.com/hashintel/hash/pull/414/commits/83ec566d92ada32890b5b34a1b1efc50aff46878